sound/soc/sof/topology.c

```diff
@@ -2146,11 +2156,34 @@ static int sof_link_hda_process(struct snd_sof_dev *sdev,
 		if (!dai->name)
 			continue;
 
-		if (strcmp(link->name, dai->name) == 0 &&
-		    dai->comp_dai.direction == direction) {
+		if (strcmp(link->name, dai->name) == 0) {
 			config->dai_index = dai->comp_dai.dai_index;
+
+			if (dai->comp_dai.direction ==
+			    SNDRV_PCM_STREAM_PLAYBACK)
+				config->hda.link_dma_ch = tx_slot;
+			else
+				config->hda.link_dma_ch = rx_slot;
+
+			/* save config in dai component */
+			dai->dai_config = kmemdup(config, size, GFP_KERNEL);
+			if (!dai->dai_config)
+				return -ENOMEM;
+
+			/* send message to DSP */
+			ret = sof_ipc_tx_message(sdev->ipc,
+						 config->hdr.cmd, config, size,
+						 &reply, sizeof(reply));
+
+			if (ret < 0) {
+				dev_err(sdev->dev, "error: failed to set DAI config for direction:%d of HDA dai %d\n",
+					dai->comp_dai.direction,
+					config->dai_index);
+
+				return ret;
+			}
+
```

Comment:
@RanderWang This looks a lot cleaner but if tx_slot and rx_slot are both non-zero, you need to make sure both are set right? But here, you only check if one is found.

Reply:
yes, both are set if non-zero. I get your idea